ssg/templates.py

```diff
@@ -0,0 +1,193 @@
+    "grub2_bootloader_argument": None,
+    "kernel_module_disabled": None,
+    "mount": None,
+    "mount_option": None,
+    "mount_option_remote_filesystems": None,
+    "mount_option_removable_partitions": None,
+    "mount_option_var": None,
+    "ocp_service_runtime_config": None,
+    "package_installed": package_installed,
+    "package_removed": None,
+    "permissions": None,
+    "sebool": None,
+    "sebool_var": None,
+    "service_disabled": None,
+    "service_enabled": None,
+    "sshd_lineinfile": None,
+    "sysctl": None,
+    "sysctl_ipv6": None,
+    "sysctl_runtime": None,
+    "sysctl_runtime_var": None,
+    "sysctl_static": None,
+    "sysctl_static_var": None,
+    "sysctl_var": None,
+    "timer_enabled": None,
+}
+
+
+class Builder(object):
+    """
+    Class for building all templated content for a given product.
+
+    To generate content from templates, pass the env_yaml, path to the
+    directory with resolved rule YAMLs, path to the directory that contains
+    templates, path to the output directory for checks and a path to the
+    output directory for remediations into the constructor. Then, call the
+    method build() to perform a build.
+    """
+    def __init__(
+            self, env_yaml, resolved_rules_dir, templates_dir,
+            remediations_dir, checks_dir):
+        self.env_yaml = env_yaml
+        self.resolved_rules_dir = resolved_rules_dir
+        self.templates_dir = templates_dir
+        self.remediations_dir = remediations_dir
+        self.checks_dir = checks_dir
+        self.output_dirs = dict()
+        for lang in languages:
+            if lang == "oval":
+                # OVAL checks need to be put to a different directory because
+                # they are processed differently than remediations later in the
+                # build process
+                output_dir = self.checks_dir
+            else:
+                output_dir = self.remediations_dir
+            dir_ = os.path.join(output_dir, lang)
+            self.output_dirs[lang] = dir_
+
+    def preprocess_data(self, template, raw_parameters):
+        """
+        Processes template data using a callback before the data will be
+        substituted into the Jinja template.
+        """
+        template_func = templates[template]
+        if template_func is not None:
+            return template_func(raw_parameters)
+        else:
+            return raw_parameters
+
+    def build_lang(self, rule, lang, jinja_dict):
+        """
+        Builds templated content for a given rule for a given language,
+        based on parameters obtained from template_data key in rule.yml.
+        Writes the output to the correct build directories.
+        """
+        template_file_name = "template_{0}_{1}".format(
+            lang.upper(), rule.template)
+        template_file_path = os.path.join(
+            self.templates_dir, template_file_name)
+        if not os.path.exists(template_file_path):
```

Comment:
I suggest to add an error message here.

Reply:
It will not work now. By default, we want to try to generate every type of content. Some templates don't have an Anaconda or Puppet option, so it's expected that the file doesn't exist.

I think if we want a meaningful error message here we would have to explicitly enumerate which templates are available for each template_name.